File: BinaryTreePostorderTraversal.java
// Stack, Tree, Depth-First Search, Binary Tree
// Facebook 2 Amazon 2 Apple 2 TuSimple 2 Bloomberg 2

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

import java.util.*;

// left-right-root
class BinaryTreePostorderTraversal {

  public List<Integer> postorderTraversalRecursion(TreeNode root) {
    List<Integer> result = new ArrayList<>();
    this.recursionHelper(root, result);

    return result;
  }

  private void recursionHelper(TreeNode root, List<Integer> result) {
    if (root == null) {
      return;
    }

    this.recursionHelper(root.left, result);
    this.recursionHelper(root.right, result);
    result.add(root.val);
  }

  public List<Integer> postorderTraversalDivideConquer(TreeNode root) {
    List<Integer> result = new ArrayList<>();
    if (root == null) {
      return result;
    }

    List<Integer> leftResults = this.postorderTraversalDivideConquer(root.left);
    List<Integer> rightResults =
      this.postorderTraversalDivideConquer(root.right);

    result.addAll(leftResults);
    result.addAll(rightResults);
    result.add(root.val);

    return result;
  }

  public List<Integer> postorderTraversal(TreeNode root) {
    List<Integer> result = new ArrayList<>();
  }
}

File: P0/BinaryTreePostorderTraversal.py
# Stack, Tree, Depth-First Search, Binary Tree
# Facebook 2 Amazon 2 Apple 2 TuSimple 2 Bloomberg 2
# https://leetcode.com/problems/binary-tree-postorder-traversal/


from collections import deque
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class BinaryTreePostorderTraversal:
    def postorderTraversal(self, root):
        result = list()
        stack = deque()
        stack.append((root, False))  # (node, visited)
        while stack:
            node, visited = stack.pop()
            if node:
                if visited:
                    # add to result if visited
                    result.append(node.val)
                else:
                    # post-order
                    stack.append((node, True))
                    stack.append((node.right, False))
                    stack.append((node.left, False))

        return result


# recursion
class BinaryTreePostorderTraversal2:
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        result = list()
        if not root:
            return result

        self.postorderTraversalHelper(root, result)

        return result

    def postorderTraversalHelper(
        self, root: Optional[TreeNode], result: List[int]
    ) -> None:
        if not root:
            return

        self.postorderTraversalHelper(root.left, result)
        self.postorderTraversalHelper(root.right, result)
        result.append(root.val)


# divide and conquer
class BinaryTreePostorderTraversal3:
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        result = list()
        if not root:
            return result

        left_result = self.postorderTraversal(root.left)
        right_result = self.postorderTraversal(root.right)
        result.extend(left_result)
        result.extend(right_result)
        result.append(root.val)

        return result
